### DISCUSSION TITLE
M2M read filter trying to show unallowed items

### DISCUSSION THREAD

**Discussion by Pachat** [directus/directus#15997] [Archive]
Between two collections, I have an M2M field called `Roles`
The second collection has read filters depending on the Directus Role
Some Directus Roles can add M2M fields that other Directus Roles are not allowed to see.

When showing an item of the first collection, Directus tries to shows all the `Roles`
It raises an exception first `"You don't have permission to access this."` 

![Exception](https://user-images.githubusercontent.com/13746133/195765811-0bf53552-3f7f-4992-a160-666300eb6010.png)

and then it shows those not allowed with `--` 

![Roles Port 1  --](https://user-images.githubusercontent.com/13746133/195765208-f9c9ce58-0e45-496a-9252-efb317fc2e30.png)

Not very friendly for the end user: he wants to get information and gets a not allowed message!
What is the use to display the `--` representing what should not be seen?
I would not like the user of a role to know that there is some information coming from other roles hidden to him.

It would be better instead just to not retrieve what should not

![Roles Port 1  -- 1](https://user-images.githubusercontent.com/13746133/195766581-2986939f-2a7d-448c-a175-b4a83739f778.png)

and just to show the items allowed by the read filter on the second collection for that role.
Is there a way to achieve this on the present version of Directus?

**Comment by azrikahar** [comment]
> The second collection has read filters depending on the Directus Role

> and then it shows those not allowed with --

I believe this is technically correct, as now the user are not allowed to view items in the second collection, but they still can view all items in the _junction_ collection for M2M. The similar read filter should also be added to the junction collection for it to work as your expectation 👍 Note that your current expectation of only setting read permission to the second collection would work for O2M (no junction), but not for M2M/M2A (with junction). 

Here's a quick example that shows before and after adding the custom read permission to the junction collection:

https://user-images.githubusercontent.com/42867097/195782844-d6d69b52-5407-4a7e-9f03-4a43fee893e4.mp4

**Reply by Pachat** [reply]
You are totally right! Thank you for the clear explanation and your time :)
